internal/pkg/docker/scheduler/scheduler.go

```diff
@@ -0,0 +1,312 @@
+		}
+
+		if err := s.stopTask(ctx, s.curTask); err != nil {
+			return fmt.Errorf("stop existing task: %w", err)
+		}
+	}
+
+	s.curTask = r.task
+	for name, ctr := range r.task.Containers {
+		name, ctr := name, ctr
+		s.run(taskID, s.containerRunOptions(name, ctr))
+	}
+
+	return nil
+}
+
+// Stop stops the current running task containers and the Scheduler. Stop is
+// idempotent and safe to call multiple times. Calls to RunTask() after calling Stop
+// do nothing.
+func (s *Scheduler) Stop() {
+	s.stopOnce.Do(func() {
+		close(s.stopped)
+		s.actions <- &stopAction{}
+	})
+}
+
+type stopAction struct{}
+
+func (a *stopAction) Do(s *Scheduler) error {
+	defer s.wg.Done()  // for the scheduler
+	s.curTaskID.Add(1) // ignore runtime errors
+
+	// collect errors since we want to try to clean up everything we can
+	var errs []error
+	if err := s.stopTask(context.Background(), s.curTask); err != nil {
+		errs = append(errs, err)
+	}
+
+	// stop pause container
+	if err := s.docker.Stop(context.Background(), s.containerID("pause")); err != nil {
+		errs = append(errs, fmt.Errorf("stop %q: %w", "pause", err))
+	}
+
+	return errors.Join(errs...)
+}
+
+// stopTask calls `docker stop` for all containers defined by task.
+func (s *Scheduler) stopTask(ctx context.Context, task Task) error {
+	if len(task.Containers) == 0 {
+		return nil
+	}
+
+	errCh := make(chan error)
+	for name := range task.Containers {
+		name := name
+		go func() {
+			if err := s.docker.Stop(ctx, s.containerID(name)); err != nil {
+				errCh <- fmt.Errorf("stop %q: %w", name, err)
+				return
+			}
+			errCh <- nil
+		}()
+	}
+
+	var errs []error
+	for err := range errCh {
+		errs = append(errs, err)
+		if len(errs) == len(task.Containers) {
+			break
+		}
+	}
+
+	return errors.Join(errs...)
+}
+
+// waitForContainerToStart blocks until the container specified by id starts.
+func (s *Scheduler) waitForContainerToStart(ctx context.Context, id string) error {
+	for {
+		isRunning, err := s.docker.IsContainerRunning(ctx, id)
+		switch {
+		case err != nil:
+			return fmt.Errorf("check if %q is running: %w", id, err)
+		case isRunning:
+			return nil
+		}
+
+		select {
+		case <-time.After(1 * time.Second):
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+}
+
+// containerID returns the full ID for a container with name run by s.
+func (s *Scheduler) containerID(name string) string {
+	return s.idPrefix + name
+}
+
+// Task defines a set of Containers to be run together.
+// Containers within a Task can talk to each other on localhost
+// and are stopped and started as a group.
+type Task struct {
+	Containers map[string]ContainerDefinition
+}
+
+// ContainerDefinition defines information necessary to run a container.
+type ContainerDefinition struct {
+	ImageURI string
+	EnvVars  map[string]string
+	Secrets  map[string]string
+	Ports    map[string]string // host port -> container port
+}
+
+// pauseRunOptions returns RunOptions for the pause container for t.
```

Comment:
maybe more clarity on what it means by "pause container for t"?

Reply:
added a small comment, hope it helps clarify!